File: clients/lumiverb-app/Sources/LumiverbKit/Sources/LumiverbKit/Models/PathFilter.swift
import Foundation

/// Path filter rule from the server.
public struct PathFilterRule: Decodable, Sendable {
    public let pattern: String
    public let filterType: String  // "include" or "exclude"
}

/// Merged path filter that applies tenant + library rules.
///
/// Priority (highest first):
/// 1. Library exclude → BLOCKED
/// 2. Library include → ALLOWED
/// 3. Tenant exclude → BLOCKED
/// 4. Tenant includes exist but no match → BLOCKED
/// 5. Default → ALLOWED
public struct PathFilter: Sendable {
    public let tenantIncludes: [String]
    public let tenantExcludes: [String]
    public let libraryIncludes: [String]
    public let libraryExcludes: [String]

    public init(
        tenantRules: [PathFilterRule] = [],
        libraryRules: [PathFilterRule] = []
    ) {
        self.tenantIncludes = tenantRules.filter { $0.filterType == "include" }.map(\.pattern)
        self.tenantExcludes = tenantRules.filter { $0.filterType == "exclude" }.map(\.pattern)
        self.libraryIncludes = libraryRules.filter { $0.filterType == "include" }.map(\.pattern)
        self.libraryExcludes = libraryRules.filter { $0.filterType == "exclude" }.map(\.pattern)
    }

    /// Initialize from server filter responses.
    public init(
        tenant: TenantFilterDefaultsResponse,
        library: LibraryFiltersResponse
    ) {
        self.tenantIncludes = tenant.includes.map(\.pattern)
        self.tenantExcludes = tenant.excludes.map(\.pattern)
        self.libraryIncludes = library.includes.map(\.pattern)
        self.libraryExcludes = library.excludes.map(\.pattern)
    }

    /// Check if a relative path is allowed by the filter rules.
    public func isAllowed(_ relPath: String) -> Bool {
        let normalized = relPath.replacingOccurrences(of: "\\", with: "/")

        // 1. Library exclude — highest priority
        for pattern in libraryExcludes {
            if globMatch(normalized, pattern: pattern) { return false }
        }

        // 2. Library include — overrides tenant
        for pattern in libraryIncludes {
            if globMatch(normalized, pattern: pattern) { return true }
        }

        // 3. Tenant exclude
        for pattern in tenantExcludes {
            if globMatch(normalized, pattern: pattern) { return false }
        }

        // 4. Tenant includes exist but no match
        if !tenantIncludes.isEmpty {
            for pattern in tenantIncludes {
                if globMatch(normalized, pattern: pattern) { return true }
            }
            return false
        }

        // 5. Default — allowed
        return true
    }

    /// Case-insensitive glob matching supporting *, ?, and **.
    ///
    /// - `*` matches any characters except `/`
    /// - `?` matches any single character except `/`
    /// - `**` matches any characters including `/` (zero or more path segments)
    private func globMatch(_ path: String, pattern: String) -> Bool {
        let p = pattern.lowercased()
        let s = path.lowercased()
        return globMatchRecursive(
            Array(s.unicodeScalars),
            Array(p.unicodeScalars),
            si: 0, pi: 0
        )
    }

    private func globMatchRecursive(
        _ s: [Unicode.Scalar], _ p: [Unicode.Scalar],
        si: Int, pi: Int
    ) -> Bool {
        var si = si
        var pi = pi

        while pi < p.count {
            if pi + 1 < p.count && p[pi] == "*" && p[pi + 1] == "*" {
                // Skip consecutive * characters
                pi += 2
                // Skip trailing /
                if pi < p.count && p[pi] == "/" { pi += 1 }
                // ** at end matches everything
                if pi >= p.count { return true }
                // Try matching remainder from every position
                for i in si...s.count {
                    if globMatchRecursive(s, p, si: i, pi: pi) {
                        return true
                    }
                }
                return false
            } else if p[pi] == "*" {
                pi += 1
                // * matches everything except /
                if pi >= p.count {
                    // * at end — match if no more slashes
                    return !s[si...].contains("/")
                }
                for i in si...s.count {
                    if i < s.count && s[i] == "/" {
                        // * doesn't cross /
                        return globMatchRecursive(s, p, si: i, pi: pi)
                    }
                    if globMatchRecursive(s, p, si: i, pi: pi) {
                        return true
                    }
                }
                return false
            } else if p[pi] == "?" {
                if si >= s.count || s[si] == "/" { return false }
                si += 1
                pi += 1
            } else {
                if si >= s.count || s[si] != p[pi] { return false }
                si += 1
                pi += 1
            }
        }

        return si >= s.count
    }
}

File: src/shared/path_filter.py
"""Shared path filter evaluation for library ingest. Pure functions, no I/O."""

from __future__ import annotations

import fnmatch
from dataclasses import dataclass
from typing import Literal


@dataclass
class PathFilter:
    type: Literal["include", "exclude"]
    pattern: str


def _glob_match(pattern: str, path: str) -> bool:
    """
    Case-insensitive glob match supporting ** for cross-segment matching.
    Both pattern and path are lowercased before comparison.
    * and ? match within a single path segment; ** matches zero or more segments.
    """
    pl = pattern.lower().replace("\\", "/")
    path_l = path.lower().replace("\\", "/")
    # Preserve empty path as single empty segment for edge cases
    path_segs = path_l.split("/") if path_l else [""]
    pattern_segs = pl.split("/")

    def match_segment(pat: str, seg: str) -> bool:
        """Single segment: fnmatch; ? and * don't cross /."""
        return fnmatch.fnmatch(seg, pat)

    def match_from(pi: int, pseg_i: int) -> bool:
        """Match path_segs[pi:] against pattern_segs[pseg_i:]. ** consumes 0+ path segments."""
        if pseg_i >= len(pattern_segs):
            return pi >= len(path_segs) or (pi == 0 and path_segs == [""])
        if pi >= len(path_segs) and path_segs != [""]:
            return all(s == "**" for s in pattern_segs[pseg_i:])
        pseg = pattern_segs[pseg_i]
        if pseg == "**":
            # Consume 0 or more path segments
            max_skip = len(path_segs) - pi + 1
            for skip in range(max_skip):
                if match_from(pi + skip, pseg_i + 1):
                    return True
            return False
        if path_segs == [""]:
            return False
        if not match_segment(pseg, path_segs[pi]):
            return False
        return match_from(pi + 1, pseg_i + 1)

    return match_from(0, 0)


def is_path_included(rel_path: str, filters: list[PathFilter]) -> bool:
    """
    Return True if rel_path passes the include/exclude filter set.

    Rules:
    - If no include filters exist, all paths start as included.
    - If any include filters exist, path must match at least one to remain included.
    - Any exclude filter match removes the path from the set.
    - Matching is case-insensitive; ** matches across path segments.
    """
    rel_path_norm = rel_path.replace("\\", "/")
    includes = [f for f in filters if f.type == "include"]
    excludes = [f for f in filters if f.type == "exclude"]

    if includes:
        if not any(_glob_match(f.pattern, rel_path_norm) for f in includes):
            return False
    for ex in excludes:
        if _glob_match(ex.pattern, rel_path_norm):
            return False
    return True


def is_path_included_merged(
    rel_path: str,
    tenant_filters: list[PathFilter],
    library_filters: list[PathFilter],
) -> bool:
    """Evaluate path against merged tenant + library filters.

    Priority rules:
    1. Library exclude matches → BLOCKED (absolute, highest priority)
    2. Library include matches → ALLOWED (overrides tenant restrictions)
    3. Tenant exclude matches → BLOCKED
    4. Tenant includes exist but path doesn't match any → BLOCKED
    5. Default → ALLOWED
    """
    rel_path_norm = rel_path.replace("\\", "/")

    lib_includes = [f for f in library_filters if f.type == "include"]
    lib_excludes = [f for f in library_filters if f.type == "exclude"]
    tenant_includes = [f for f in tenant_filters if f.type == "include"]
    tenant_excludes = [f for f in tenant_filters if f.type == "exclude"]

    # Rule 1: Library exclude → blocked
    if any(_glob_match(f.pattern, rel_path_norm) for f in lib_excludes):
        return False

    # Rule 2: Library include → allowed (overrides tenant restrictions)
    if lib_includes and any(_glob_match(f.pattern, rel_path_norm) for f in lib_includes):
        return True

    # Rule 3: Tenant exclude → blocked
    if any(_glob_match(f.pattern, rel_path_norm) for f in tenant_excludes):
        return False

    # Rule 4: Tenant includes exist but path doesn't match → blocked
    if tenant_includes and not any(_glob_match(f.pattern, rel_path_norm) for f in tenant_includes):
        return False

    # Rule 5: Default → allowed
    return True


def validate_pattern(pattern: str) -> str:
    """
    Validate and return the pattern, or raise ValueError if invalid.
    Rejects patterns containing '..' or null bytes.
    """
    if "\x00" in pattern:
        raise ValueError("Pattern must not contain null bytes")
    if ".." in pattern:
        raise ValueError("Pattern must not contain '..'")
    return pattern
